### PR TITLE
Add LLAP ReadSupport

### PR DESCRIPTION
**Pull request by EricWohlstadter**
## What changes were proposed in this pull request?

Add impls. for the ReadSupport interfaces: DataSourceV2, DataSourceReader, DataSourceReaderFactory, and DataReader.

Refactor some code into Util classes.

Move some non-static methods in LlapRelation classes to DataSourceV2 classes.

Add some Mock classes to support UT.

## How was this patch tested?

UT

### CODE REVIEW

**Comment by EricWohlstadter** [comment]
Test failures are expected.
Tests pass on my local with Hive branch-3.

**Comment by dongjoon-hyun** [comment]
For the build failure, which Hive Apache JIRA is adding `org.apache.hadoop.hive.ql.io.arrow` then?
Should we use Apache Hive snapshot instead of internal repo?

**Comment by EricWohlstadter** [comment]
@dongjoon-hyun 

To build, we need:
1. HIVE-19306 (introduces ArrowWrapperWritable and RootAllocatorFactory)
2. HIVE-19495 (refactors HIVE-19306)
3. HIVE-19308 (introduces LlapArrowBatchRecordReader and modifies LlapBaseInputFormat)

**Comment by dongjoon-hyun** [comment]
Could you build, verify, and merge this manually, @HyukjinKwon ?

**Comment by HyukjinKwon** [comment]
Sure.

**Comment by HyukjinKwon** [comment]
Hmmm, @EricWohlstadter, were you able to build  with Hive 3.1.0-SNAPSHOT? Seems I am facing dependency issues.

**Comment by HyukjinKwon** [comment]
@EricWohlstadter, are you able to build this against Hive 3.1.0-SNAPSHOT? Seems I can't even with locally installed Hive one.

**Comment by EricWohlstadter** [comment]
@HyukjinKwon 

What build error are you seeing?
I am able to pass build/test locally after doing a local mvn install.
I'm using:
```
build/sbt clean assembly -Drepourl=http://nexus-private.hortonworks.com/nexus/content/groups/public/ -Dhadoop.version=3.0.0.3.0.0.0-SNAPSHOT -Dhive.version=3.1.0-SNAPSHOT
```

**Comment by HyukjinKwon** [comment]
@EricWohlstadter, before merging this, mind if I ask to fix some style nits by running `build/scalastyle`?

File: src/main/java/com/hortonworks/spark/sql/hive/llap/util/HiveQlUtil.java [hunk 1]
```diff
@@ -86,4 +103,8 @@ public static String createTablePrelude(String database, String table, boolean i
     private static String orBlank(boolean useText, String text) {
         return (useText ? text : "");
     }
+
+    public static String randomAlias() {
+        return "q_" + UUID.randomUUID().toString().replaceAll("[^A-Za-z0-9 ]", "");
```
Comment (HyukjinKwon) [inline src/main/java/com/hortonworks/spark/sql/hive/llap/util/HiveQlUtil.java]: how about just random number?

File: src/main/scala/com/hortonworks/spark/sql/hive/llap/FilterPushdown.scala [hunk 2]
```diff
@@ -26,6 +26,25 @@ private[llap] object FilterPushdown extends Object {
...
+          GreaterThanOrEqual (_, _) |
+          In (_, _) |
+          StringStartsWith (_, _) |
+          StringEndsWith (_, _) |
+          StringContains (_, _) |
+          IsNull (_) |
+          IsNotNull (_) |
+          And (_, _) |
+          Or (_, _) |
+          Not (_) => true
```
Comment (HyukjinKwon) [inline src/main/scala/com/hortonworks/spark/sql/hive/llap/FilterPushdown.scala]: @EricWohlstadter, seems it's not going to support nested filters.

**Comment by HyukjinKwon** [comment]
Actually, will take another look.

File: src/main/scala/com/hortonworks/spark/sql/hive/llap/FilterPushdown.scala (hunk 2 above)
Comment (HyukjinKwon) [reply src/main/scala/com/hortonworks/spark/sql/hive/llap/FilterPushdown.scala]: Can we maybe, instead, use `buildFilterExpression(...).isDefined`?

Comment (EricWohlstadter) [reply src/main/scala/com/hortonworks/spark/sql/hive/llap/FilterPushdown.scala]: @HyukjinKwon 

Good idea. I made that change.

**Pull request merged**